src/core/Tree/Tree.tsx

```diff
@@ -0,0 +1,228 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps, useTheme } from '../utils';
+import '@itwin/itwinui-css/css/tree.css';
+import cx from 'classnames';
+
+export const TreeContext = React.createContext<
+  | {
+      selectedNodes?: Array<string>;
+      setSelectedNode?: (
+        node: Array<string> | ((prev: Array<string>) => Array<string>),
+      ) => void;
+      expandedNodes?: Array<string>;
+      setExpandedNodes?: (
+        node: Array<string> | ((prev: Array<string>) => Array<string>),
+      ) => void;
+      nodeDepth?: number;
+      parentNode?: FlatNode;
+    }
+  | undefined
+>(undefined);
+
+export type TreeData = {
+  subnodes: Array<TreeData>;
+  nodeId: string;
+  label?: string;
+  subLabel?: string;
+  isExpanded?: boolean;
+  isDisabled?: boolean;
+};
+
+type FlatNode = {
+  data: TreeData;
+  id: string;
+  depth: number;
+  subNodeIds: string[];
+  parent?: FlatNode;
+};
+
+export type TreeProps = {
```

Comment:
I think Saally you will hate me for this 😀 I am thinking maybe we could get best of what others have by having a prop `getNode: (node: TreeData, index: number) => TreeData` and we could call in in `flatNodesList`. This way users could set `isExpanded` and `isSelected` whatever way they want: have separate lists where expanded node ids are saved or saving inside their node data properties. It may also work for Robertas too e.g. passing `data={new Array(500)}` and then according `getNode`.

Reply:
@mayank99 @veekeys would be nice to hear your opinion.

Reply:
I find the current api already confusing. There is `NodeData` and `FlatNode`, which seem to be inexplicably separated out.

If `getNode` is added, it would create more confusion with `nodeRenderer` which already has some handlers. How does `getNode` help exactly?

I think it would make sense to separate out the state from the UI and then expose two props: `initialState` and `stateReducer`. And we can keep the `nodeRenderer` but it should only be used for the UI, and not have any handlers.

Reply:
But how users are going to pass extended nodes to `initialState`? Than we might have to make `id` prop required in their data.

Reply:
I added `getNode` to Tree props, so user can set isExpanded. I also combined `NodeData` and `FlatNode` since I think both are not needed anymore.